File: cloud.js
// ============================================================
//  Casri POS — cloud sync
// ------------------------------------------------------------
//  The app is bundled offline and runs from a file:// URL inside the APK.
//  The Firebase JS SDK is unreliable from that origin (and would add ~200KB to
//  bundle), so this talks to Firebase over its plain REST APIs with fetch():
//
//    auth  → identitytoolkit.googleapis.com   (email + password)
//    data  → firestore.googleapis.com/v1      (Bearer idToken)
//
//  Model: one Firestore document per storage key, under the signed-in user:
//      casripos/{uid}/keys/{pos_prod | pos_sales | …}
//  Each doc holds the same JSON string the app already keeps in localStorage,
//  plus a millisecond timestamp. Newest timestamp wins — the same last-writer
//  -wins rule Isguul uses, which is safe because one shop edits at a time.
//
//  Per-key documents (rather than one big blob) keep each key under Firestore's
//  1 MB document limit and mean a sales push doesn't rewrite the product list.
// ============================================================

var CLOUD = {
  cfg: null,        // {apiKey, projectId}
  email: "",
  uid: "",
  idToken: "",
  refreshToken: "",
  tokenAt: 0,
  on: false,        // signed in and syncing
  busy: false,
  lastSync: 0,
  lastError: "",
  _timer: null,
  _pending: {}
};

var CLOUD_KEYS = ["pos_biz_list", "pos_current_biz", "pos_prod", "pos_sales",
                  "pos_inv", "pos_acc", "pos_fx"];

// ── config + session persistence ────────────────────────────
function cloudLoad() {
  try {
    var s = JSON.parse(localStorage.getItem("pos_cloud") || "null");
    if (s) {
      CLOUD.cfg = s.cfg || null;
      CLOUD.email = s.email || "";
      CLOUD.uid = s.uid || "";
      CLOUD.refreshToken = s.refreshToken || "";
      CLOUD.lastSync = s.lastSync || 0;
    }
  } catch (e) {}
  if (!CLOUD.cfg && typeof window !== "undefined" && window.BUNDLED_CLOUD_CFG) {
    CLOUD.cfg = window.BUNDLED_CLOUD_CFG;
  }
}
function cloudSave() {
  try {
    localStorage.setItem("pos_cloud", JSON.stringify({
      cfg: CLOUD.cfg, email: CLOUD.email, uid: CLOUD.uid,
      refreshToken: CLOUD.refreshToken, lastSync: CLOUD.lastSync
    }));
  } catch (e) {}
}

function _cloudOnline() {
  return !(typeof navigator !== "undefined" && navigator && navigator.onLine === false);
}

// ── auth ────────────────────────────────────────────────────
function _authURL(method) {
  return "https://identitytoolkit.googleapis.com/v1/accounts:" + method +
         "?key=" + encodeURIComponent(CLOUD.cfg.apiKey);
}
function _cloudAuth(method, email, password) {
  return fetch(_authURL(method), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: email, password: password, returnSecureToken: true })
  }).then(function (r) {
    return r.json().then(function (j) {
      if (!r.ok) {
        var code = (j && j.error && j.error.message) || "AUTH_FAILED";
        throw new Error(code);
      }
      return j;
    });
  });
}
function _applySession(j) {
  CLOUD.idToken = j.idToken;
  CLOUD.refreshToken = j.refreshToken;
  CLOUD.uid = j.localId;
  CLOUD.tokenAt = Date.now();
  CLOUD.on = true;
  cloudSave();
}
// idTokens last an hour; refresh a little early.
function _cloudFreshToken() {
  if (CLOUD.idToken && (Date.now() - CLOUD.tokenAt) < 50 * 60 * 1000) {
    return Promise.resolve(CLOUD.idToken);
  }
  if (!CLOUD.refreshToken) return Promise.reject(new Error("NO_SESSION"));
  return fetch("https://securetoken.googleapis.com/v1/token?key=" + encodeURIComponent(CLOUD.cfg.apiKey), {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: "grant_type=refresh_token&refresh_token=" + encodeURIComponent(CLOUD.refreshToken)
  }).then(function (r) { return r.json(); }).then(function (j) {
    if (!j.id_token) throw new Error("REFRESH_FAILED");
    CLOUD.idToken = j.id_token;
    CLOUD.refreshToken = j.refresh_token || CLOUD.refreshToken;
    CLOUD.tokenAt = Date.now();
    cloudSave();
    return CLOUD.idToken;
  });
}

// Friendly text for Firebase's shouty error codes.
function _cloudErrText(code) {
  var m = {
    EMAIL_NOT_FOUND:            T("No account with that email", "Email-kaas akoon ma laha"),
    INVALID_PASSWORD:           T("Wrong password", "Furaha waa khalad"),
    INVALID_LOGIN_CREDENTIALS:  T("Wrong email or password", "Email ama furaha waa khalad"),
    EMAIL_EXISTS:               T("That email already has an account", "Email-kaas akoon ayuu leeyahay"),
    WEAK_PASSWORD:              T("Password must be at least 6 characters", "Furuhu waa inuu ka koobnaadaa 6 xaraf"),
    INVALID_EMAIL:              T("That email address isn't valid", "Email-kaasi sax ma aha"),
    OPERATION_NOT_ALLOWED:      T("Email sign-in is not enabled in Firebase", "Email-ka lagama furin Firebase"),
    NO_SESSION:                 T("Sign in first", "Marka hore gal"),
    TOO_MANY_ATTEMPTS_TRY_LATER: T("Too many attempts — try later", "Isku dayo badan — hadhow isku day")
  };
  if (m[code]) return m[code];
  if (String(code).indexOf("Failed to fetch") >= 0 || String(code).indexOf("NetworkError") >= 0) {
    return T("No internet connection", "Internet ma jiro");
  }
  return String(code || "Error");
}

// ── firestore REST ──────────────────────────────────────────
function _docURL(key) {
  return "https://firestore.googleapis.com/v1/projects/" + CLOUD.cfg.projectId +
         "/databases/(default)/documents/casripos/" + CLOUD.uid + "/keys/" + key;
}
function _putKey(key, token) {
  var raw = localStorage.getItem(key);
  if (raw === null) return Promise.resolve(false);
  var body = { fields: { v: { stringValue: raw }, ts: { integerValue: String(Date.now()) } } };
  return fetch(_docURL(key), {
    method: "PATCH",
    headers: { "Content-Type": "application/json", Authorization: "Bearer " + token },
    body: JSON.stringify(body)
  }).then(function (r) {
    if (!r.ok) return r.json().then(function (j) {
      throw new Error((j && j.error && j.error.message) || ("HTTP " + r.status));
    });
    return true;
  });
}
function _getKey(key, token) {
  return fetch(_docURL(key), { headers: { Authorization: "Bearer " + token } })
    .then(function (r) {
      if (r.status === 404) return null;          // nothing in the cloud yet
      if (!r.ok) return r.json().then(function (j) {
        throw new Error((j && j.error && j.error.message) || ("HTTP " + r.status));
      });
      return r.json();
    })
    .then(function (doc) {
      if (!doc || !doc.fields || !doc.fields.v) return null;
      return {
        v: doc.fields.v.stringValue,
        ts: parseInt((doc.fields.ts && doc.fields.ts.integerValue) || "0", 10)
      };
    });
}

// ── push / pull ─────────────────────────────────────────────
// Writes are batched: a burst of saves (a sale touches products AND sales)
// becomes one push a moment later instead of several.
function cloudQueue(key) {
  if (!CLOUD.on || CLOUD_KEYS.indexOf(key) < 0) return;
  CLOUD._pending[key] = true;
  if (CLOUD._timer) clearTimeout(CLOUD._timer);
  CLOUD._timer = setTimeout(cloudFlush, 1500);
}
function cloudFlush() {
  if (!CLOUD.on || CLOUD.busy || !_cloudOnline()) return Promise.resolve();
  var keys = Object.keys(CLOUD._pending);
  if (!keys.length) return Promise.resolve();
  CLOUD._pending = {};
  CLOUD.busy = true; _cloudPaint("sync");
  return _cloudFreshToken().then(function (token) {
    return Promise.all(keys.map(function (k) { return _putKey(k, token); }));
  }).then(function () {
    CLOUD.lastSync = Date.now(); CLOUD.lastError = ""; cloudSave();
    CLOUD.busy = false; _cloudPaint("ok");
  }).catch(function (e) {
    // Keep the keys queued so a later flush retries them.
    keys.forEach(function (k) { CLOUD._pending[k] = true; });
    CLOUD.lastError = _cloudErrText(e.message); CLOUD.busy = false; _cloudPaint("err");
  });
}
// Pull every key. Remote wins only when its timestamp is newer than ours.
function cloudPull(force) {
  if (!CLOUD.on || !_cloudOnline()) return Promise.resolve(0);
  CLOUD.busy = true; _cloudPaint("sync");
  var applied = 0;
  return _cloudFreshToken().then(function (token) {
    return Promise.all(CLOUD_KEYS.map(function (k) {
      return _getKey(k, token).then(function (remote) {
        if (!remote) return;
        var localTs = parseInt(localStorage.getItem("pos_ts_" + k) || "0", 10);
        if (!force && remote.ts <= localTs) return;
        if (localStorage.getItem(k) === remote.v) return;
        localStorage.setItem(k, remote.v);
        localStorage.setItem("pos_ts_" + k, String(remote.ts));
        applied++;
      });
    }));
  }).then(function () {
    CLOUD.lastSync = Date.now(); CLOUD.lastError = ""; cloudSave();
    CLOUD.busy = false; _cloudPaint("ok");
    return applied;
  }).catch(function (e) {
    CLOUD.lastError = _cloudErrText(e.message); CLOUD.busy = false; _cloudPaint("err");
    return 0;
  });
}

// What is already in the cloud for this account? Used at sign-in to decide
// direction. NEVER infer this from timestamps: a freshly installed phone has
// NEWER stamps than a PC that uploaded yesterday, so a timestamp-based guess
// would push the empty phone over the real data and destroy it.
function cloudRemoteInfo() {
  return _cloudFreshToken().then(function (token) {
    return Promise.all(CLOUD_KEYS.map(function (k) {
      return _getKey(k, token).then(function (r) { return { k: k, r: r }; });
    }));
  }).then(function (rows) {
    var info = { has: false, businesses: 0, products: 0, sales: 0 };
    rows.forEach(function (row) {
      if (!row.r || !row.r.v) return;
      var n = 0;
      try { var a = JSON.parse(row.r.v); n = Array.isArray(a) ? a.length : 0; } catch (e) {}
      if (row.k === "pos_biz_list") info.businesses = n;
      if (row.k === "pos_prod")     info.products = n;
      if (row.k === "pos_sales")    info.sales = n;
      if (n > 0) info.has = true;
    });
    return info;
  });
}
// Local counts, for the same comparison.
function cloudLocalInfo() {
  function n(k) {
    try { var a = JSON.parse(localStorage.getItem(k) || "[]"); return Array.isArray(a) ? a.length : 0; }
    catch (e) { return 0; }
  }
  return { businesses: n("pos_biz_list"), products: n("pos_prod"), sales: n("pos_sales") };
}

// ── public actions (wired to Settings) ──────────────────────
// Signs in only. The CALLER decides which way data should move, after showing
// the user what is on each side.
function cloudSignIn(email, password, isNew) {
  if (!CLOUD.cfg || !CLOUD.cfg.apiKey) {
    return Promise.reject(new Error(T("Cloud is not configured", "Cloud lama habayn")));
  }
  return _cloudAuth(isNew ? "signUp" : "signInWithPassword", email, password)
    .then(function (j) {
      CLOUD.email = email;
      _applySession(j);
      return cloudRemoteInfo();
    });
}
function cloudSignOut() {
  CLOUD.on = false; CLOUD.idToken = ""; CLOUD.refreshToken = "";
  CLOUD.uid = ""; CLOUD.email = "";
  cloudSave();
  _cloudPaint("off");
}
// Push everything now — used after signing in on a device that already has data.
function cloudPushAll() {
  if (!CLOUD.on) return Promise.resolve();
  CLOUD_KEYS.forEach(function (k) { CLOUD._pending[k] = true; });
  return cloudFlush();
}
// Restore the session on boot, then quietly pull.
function cloudBoot() {
  cloudLoad();
  if (!CLOUD.cfg || !CLOUD.refreshToken) { _cloudPaint("off"); return; }
  CLOUD.on = true;
  _cloudPaint("sync");
  _cloudFreshToken()
    .then(function () { return cloudPull(false); })
    .then(function (n) {
      if (n > 0 && typeof renderPage === "function" && typeof PAGE !== "undefined") {
        location.reload();          // data changed underneath us
      }
    })
    .catch(function (e) {
      CLOUD.lastError = _cloudErrText(e.message);
      _cloudPaint("err");
    });
}

File: flutter_app/lib/data/cloud.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

/// Casri POS — cloud sync, ported 1:1 from the web app's cloud.js so both
/// versions share ONE Firebase backend and the same documents.
///
///   auth  → identitytoolkit.googleapis.com   (email + password)
///   data  → firestore.googleapis.com/v1      (Bearer idToken)
///
/// One Firestore document per storage key, under the signed-in user:
///   casripos/{uid}/keys/{pos_prod | pos_sales | …}
/// Each doc holds the same JSON string kept on-device, plus a ms timestamp.
/// Newest timestamp wins — one shop edits at a time, so last-writer-wins is safe.

// Public web API key + project id. Safe to ship: they identify the project, they
// do not grant access. Access is Firebase Auth + Firestore rules scoped to uid.
const _apiKey = 'AIzaSyCEZxp9W7_h2Nu1qs_wiQdrbXARVb5yvg8';
const _projectId = 'isguul-togdheer';

/// The MareegTech super-admin. Signs in through the normal workspace login and
/// gets the Workspaces approval console instead of a till.
const kMasterEmail = 'admin@mareegtech.com';

const cloudKeys = [
  'pos_biz_list', 'pos_current_biz', 'pos_prod', 'pos_sales',
  'pos_inv', 'pos_acc', 'pos_fx', 'pos_recovery_email',
];

/// The MPQ plan tiers, matching the web app.
class Plan {
  final String id;
  final String label;
  final int maxProducts;
  final int registers;
  const Plan(this.id, this.label, this.maxProducts, this.registers);
}

const plans = <String, Plan>{
  'MPQ50': Plan('MPQ50', 'MPQ50 — 50 products, 1 register', 50, 1),
  'MPQ100': Plan('MPQ100', 'MPQ100 — 100 products, 2 registers', 100, 2),
  'MPQ200': Plan('MPQ200', 'MPQ200 — 200 products, 3 registers', 200, 3),
};

/// Result of inspecting either side before choosing a sync direction.
class SyncInfo {
  final bool has;
  final int businesses, products, sales;
  const SyncInfo(
      {this.has = false, this.businesses = 0, this.products = 0, this.sales = 0});
}

/// A workspace row in the approval registry.
class Workspace {
  final String uid, email, name, plan;
  final bool approved;
  final int createdAt;
  const Workspace({
    required this.uid,
    this.email = '',
    this.name = '',
    this.plan = '',
    this.approved = false,
    this.createdAt = 0,
  });
}

class Cloud extends ChangeNotifier {
  SharedPreferences? _sp;

  String email = '';
  String uid = '';
  String _idToken = '';
  String _refreshToken = '';
  int _tokenAt = 0;
  bool on = false;
  bool busy = false;
  int lastSync = 0;
  String lastError = '';

  // 'off' | 'sync' | 'ok' | 'err'
  String status = 'off';

  final Map<String, bool> _pending = {};
  Timer? _timer;

  bool get master =>
      on && email.trim().toLowerCase() == kMasterEmail.trim().toLowerCase();

  void _paint(String s) {
    status = s;
    notifyListeners();
  }

  // ── session persistence (shares the web app's `pos_cloud` shape) ──────────
  Future<void> load() async {
    _sp ??= await SharedPreferences.getInstance();
    try {
      final raw = _sp!.getString('pos_cloud');
      if (raw != null && raw.isNotEmpty) {
        final s = Map<String, dynamic>.from(jsonDecode(raw));
        email = (s['email'] ?? '').toString();
        uid = (s['uid'] ?? '').toString();
        _refreshToken = (s['refreshToken'] ?? '').toString();
        lastSync = (s['lastSync'] as num?)?.toInt() ?? 0;
      }
    } catch (_) {}
  }

  Future<void> _save() async {
    _sp ??= await SharedPreferences.getInstance();
    await _sp!.setString(
        'pos_cloud',
        jsonEncode({
          'email': email,
          'uid': uid,
          'refreshToken': _refreshToken,
          'lastSync': lastSync,
        }));
  }

  // ── auth ──────────────────────────────────────────────────────────────────
  Uri _authUrl(String method) => Uri.parse(
      'https://identitytoolkit.googleapis.com/v1/accounts:$method?key=$_apiKey');

  Future<Map<String, dynamic>> _auth(
      String method, String em, String pw) async {
    final r = await http.post(_authUrl(method),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(
            {'email': em, 'password': pw, 'returnSecureToken': true}));
    final j = Map<String, dynamic>.from(jsonDecode(r.body));
    if (r.statusCode >= 400) {
      final code = (j['error']?['message'] ?? 'AUTH_FAILED').toString();
      throw CloudError(code);
    }
    return j;
  }

  Future<void> _applySession(Map<String, dynamic> j) async {
    _idToken = (j['idToken'] ?? '').toString();
    _refreshToken = (j['refreshToken'] ?? '').toString();
    uid = (j['localId'] ?? '').toString();
    _tokenAt = DateTime.now().millisecondsSinceEpoch;
    on = true;
    await _save();
  }

  // idTokens last an hour; refresh a little early.
  Future<String> _freshToken() async {
    final now = DateTime.now().millisecondsSinceEpoch;
    if (_idToken.isNotEmpty && (now - _tokenAt) < 50 * 60 * 1000) {
      return _idToken;
    }
    if (_refreshToken.isEmpty) throw CloudError('NO_SESSION');
    final r = await http.post(
        Uri.parse('https://securetoken.googleapis.com/v1/token?key=$_apiKey'),
        headers: {'Content-Type': 'application/x-www-form-urlencoded'},
        body:
            'grant_type=refresh_token&refresh_token=${Uri.encodeComponent(_refreshToken)}');
    final j = Map<String, dynamic>.from(jsonDecode(r.body));
    if (j['id_token'] == null) throw CloudError('REFRESH_FAILED');
    _idToken = j['id_token'].toString();
    _refreshToken = (j['refresh_token'] ?? _refreshToken).toString();
    _tokenAt = DateTime.now().millisecondsSinceEpoch;
    await _save();
    return _idToken;
  }

  // ── firestore per-key docs ─────────────────────────────────────────────────
  Uri _docUrl(String key) => Uri.parse(
      'https://firestore.googleapis.com/v1/projects/$_projectId/databases/(default)/documents/casripos/$uid/keys/$key');

  Future<bool> _putKey(String key, String token) async {
    _sp ??= await SharedPreferences.getInstance();
    final raw = _sp!.getString(key);
    if (raw == null) return false;
    final body = jsonEncode({
      'fields': {
        'v': {'stringValue': raw},
        'ts': {
          'integerValue': DateTime.now().millisecondsSinceEpoch.toString()
        },
      }
    });
    final r = await http.patch(_docUrl(key),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token'
        },
        body: body);
    if (r.statusCode >= 400) throw CloudError(_httpErr(r));
    return true;
  }

  /// Returns {v, ts} or null when nothing is in the cloud yet.
  Future<Map<String, dynamic>?> _getKey(String key, String token) async {
    final r =
        await http.get(_docUrl(key), headers: {'Authorization': 'Bearer $token'});
    if (r.statusCode == 404) return null;
    if (r.statusCode >= 400) throw CloudError(_httpErr(r));
    final doc = Map<String, dynamic>.from(jsonDecode(r.body));
    final f = doc['fields'];
    if (f == null || f['v'] == null) return null;
    return {
      'v': f['v']['stringValue'],
      'ts': int.tryParse((f['ts']?['integerValue'] ?? '0').toString()) ?? 0,
    };
  }

  String _httpErr(http.Response r) {
    try {
      final j = Map<String, dynamic>.from(jsonDecode(r.body));
      return (j['error']?['message'] ?? 'HTTP ${r.statusCode}').toString();
    } catch (_) {
      return 'HTTP ${r.statusCode}';
    }
  }

  // ── push / pull ─────────────────────────────────────────────────────────────
  void queue(String key) {
    if (!on || !cloudKeys.contains(key)) return;
    _pending[key] = true;
    _timer?.cancel();
    _timer = Timer(const Duration(milliseconds: 1500), flush);
  }

  Future<void> flush() async {
    if (!on || busy) return;
    final keys = _pending.keys.toList();
    if (keys.isEmpty) return;
    _pending.clear();
    busy = true;
    _paint('sync');
    try {
      final token = await _freshToken();
      for (final k in keys) {
        await _putKey(k, token);
      }
      lastSync = DateTime.now().millisecondsSinceEpoch;
      lastError = '';
      await _save();
      busy = false;
      _paint('ok');
    } catch (e) {
      for (final k in keys) {
        _pending[k] = true;
      }
      lastError = errText(e);
      busy = false;
      _paint('err');
    }
  }

  /// Pull every key. Remote wins only when its timestamp is newer than ours.
  /// Returns how many keys were applied locally.
  Future<int> pull({bool force = false}) async {
    if (!on) return 0;
    _sp ??= await SharedPreferences.getInstance();
    busy = true;
    _paint('sync');
    var applied = 0;
    try {
      final token = await _freshToken();
      for (final k in cloudKeys) {
        final remote = await _getKey(k, token);
        if (remote == null) continue;
        final localTs = _sp!.getInt('pos_ts_$k') ?? 0;
        final rts = (remote['ts'] as int?) ?? 0;
        if (!force && rts <= localTs) continue;
        if (_sp!.getString(k) == remote['v']) continue;
        await _sp!.setString(k, remote['v'].toString());
        await _sp!.setInt('pos_ts_$k', rts);
        applied++;
      }
      lastSync = DateTime.now().millisecondsSinceEpoch;
      lastError = '';
      await _save();
      busy = false;
      _paint('ok');
      return applied;
    } catch (e) {
      lastError = errText(e);
      busy = false;
      _paint('err');
      return 0;
    }
  }

  /// Push every key now — used after signing in on a device that already has data.
  Future<void> pushAll() async {
    if (!on) return;
    for (final k in cloudKeys) {
      _pending[k] = true;
    }
    await flush();
  }

  /// What is already in the cloud for this account? Used at sign-in to choose
  /// direction. NEVER infer this from timestamps — a fresh install has newer
  /// stamps than a PC that uploaded yesterday, so a timestamp guess would push
  /// the empty device over the real data and destroy it.
  Future<SyncInfo> remoteInfo() async {
    final token = await _freshToken();
    var has = false;
    var b = 0, p = 0, s = 0;
    for (final k in cloudKeys) {
      final r = await _getKey(k, token);
      if (r == null || r['v'] == null) continue;
      var n = 0;
      try {
        final a = jsonDecode(r['v'].toString());
        n = a is List ? a.length : 0;
      } catch (_) {}
      if (k == 'pos_biz_list') b = n;
      if (k == 'pos_prod') p = n;
      if (k == 'pos_sales') s = n;
      if (n > 0) has = true;
    }
    return SyncInfo(has: has, businesses: b, products: p, sales: s);
  }

  Future<SyncInfo> localInfo() async {
    _sp ??= await SharedPreferences.getInstance();
    int n(String k) {
      try {
        final a = jsonDecode(_sp!.getString(k) ?? '[]');
        return a is List ? a.length : 0;
      } catch (_) {
        return 0;
      }
    }

    return SyncInfo(
        businesses: n('pos_biz_list'),
        products: n('pos_prod'),
        sales: n('pos_sales'));
  }

  // ── public actions ──────────────────────────────────────────────────────────
  /// Signs in (or signs up). The CALLER decides which way data should move.
  Future<SyncInfo> signIn(String em, String pw, {bool isNew = false}) async {
    final j = await _auth(isNew ? 'signUp' : 'signInWithPassword', em, pw);
    email = em;
    await _applySession(j);
    _paint('ok');
    return remoteInfo();
  }

  Future<void> signOut() async {
    on = false;
    _idToken = '';
    _refreshToken = '';
    uid = '';
    email = '';
    _pending.clear();
    await _save();
    _paint('off');
  }

  // ── workspace approval registry ─────────────────────────────────────────────
  //  casripos_workspaces/{uid}: { email, name, plan, approved, createdAt }.
  //  A client reads/writes only its OWN doc; the master lists all and flips
  //  `approved`. Holds NO shop data — products/sales stay private to each uid.
  Uri _wsUrl([String? u]) => Uri.parse(
      'https://firestore.googleapis.com/v1/projects/$_projectId/databases/(default)/documents/casripos_workspaces/${u ?? uid}');

  Future<void> registerWorkspace(String name, String plan) async {
    final token = await _freshToken();
    final body = jsonEncode({
      'fields': {
        'email': {'stringValue': email},
        'name': {'stringValue': name},
        'plan': {'stringValue': plan},
        'approved': {'booleanValue': false},
        'createdAt': {
          'integerValue': DateTime.now().millisecondsSinceEpoch.toString()
        },
      }
    });
    final r = await http.patch(_wsUrl(),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token'
        },
        body: body);
    if (r.statusCode >= 400) throw CloudError(_httpErr(r));
  }

  /// This account's approval status.
  Future<Workspace?> workspaceStatus() async {
    final token = await _freshToken();
    final r = await http.get(_wsUrl(), headers: {'Authorization': 'Bearer $token'});
    if (r.statusCode == 404) return null;
    if (r.statusCode >= 400) throw CloudError(_httpErr(r));
    final doc = Map<String, dynamic>.from(jsonDecode(r.body));
    final f = doc['fields'] ?? {};
    return Workspace(
      uid: uid,
      email: (f['email']?['stringValue'] ?? '').toString(),
      name: (f['name']?['stringValue'] ?? '').toString(),
      plan: (f['plan']?['stringValue'] ?? '').toString(),
      approved: f['approved']?['booleanValue'] == true,
    );
  }

  /// Master: list every workspace.
  Future<List<Workspace>> listWorkspaces() async {
    final token = await _freshToken();
    final url = Uri.parse(
        'https://firestore.googleapis.com/v1/projects/$_projectId/databases/(default)/documents/casripos_workspaces?pageSize=300');
    final r = await http.get(url, headers: {'Authorization': 'Bearer $token'});
    if (r.statusCode >= 400) throw CloudError(_httpErr(r));
    final j = Map<String, dynamic>.from(jsonDecode(r.body));
    final docs = (j['documents'] as List?) ?? [];
    return docs.map((d) {
      final doc = Map<String, dynamic>.from(d);
      final f = doc['fields'] ?? {};
      final nm = (doc['name'] ?? '').toString();
      return Workspace(
        uid: nm.substring(nm.lastIndexOf('/') + 1),
        email: (f['email']?['stringValue'] ?? '').toString(),
        name: (f['name']?['stringValue'] ?? '').toString(),
        plan: (f['plan']?['stringValue'] ?? '').toString(),
        approved: f['approved']?['booleanValue'] == true,
        createdAt:
            int.tryParse((f['createdAt']?['integerValue'] ?? '0').toString()) ??
                0,
      );
    }).toList();
  }

  /// Master: approve (or revoke) a workspace — patches only `approved`.
  Future<void> approveWorkspace(String u, bool approved) async {
    final token = await _freshToken();
    final url = Uri.parse(
        'https://firestore.googleapis.com/v1/projects/$_projectId/databases/(default)/documents/casripos_workspaces/$u?updateMask.fieldPaths=approved');
    final r = await http.patch(url,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token'
        },
        body: jsonEncode({
          'fields': {
            'approved': {'booleanValue': approved}
          }
        }));
    if (r.statusCode >= 400) throw CloudError(_httpErr(r));
  }

  /// Restore the session on boot and return keys applied by a quiet pull.
  Future<int> boot() async {
    await load();
    if (_refreshToken.isEmpty) {
      _paint('off');
      return 0;
    }
    on = true;
    _paint('sync');
    try {
      await _freshToken();
      return await pull(force: false);
    } catch (e) {
      lastError = errText(e);
      _paint('err');
      return 0;
    }
  }

  // Friendly text for Firebase's shouty error codes.
  static String errText(Object e) {
    final code = e is CloudError ? e.code : e.toString();
    const m = {
      'EMAIL_NOT_FOUND': 'No account with that email',
      'INVALID_PASSWORD': 'Wrong password',
      'INVALID_LOGIN_CREDENTIALS': 'Wrong email or password',
      'EMAIL_EXISTS': 'That email already has an account',
      'WEAK_PASSWORD': 'Password must be at least 6 characters',
      'INVALID_EMAIL': "That email address isn't valid",
      'OPERATION_NOT_ALLOWED': 'Email sign-in is not enabled',
      'NO_SESSION': 'Sign in first',
      'TOO_MANY_ATTEMPTS_TRY_LATER': 'Too many attempts — try later',
    };
    for (final k in m.keys) {
      if (code.contains(k)) return m[k]!;
    }
    if (code.contains('SocketException') ||
        code.contains('Failed host lookup') ||
        code.contains('ClientException')) {
      return 'No internet connection';
    }
    return code;
  }
}

class CloudError implements Exception {
  final String code;
  CloudError(this.code);
  @override
  String toString() => code;
}

/// The one cloud client, shared by the store (for sync) and the UI (for status).
final cloud = Cloud();
